src/transform/src/normalize_lets.rs

```diff
@@ -132,32 +132,26 @@ impl NormalizeLets {
 
         // A final bottom-up traversal to normalize the shape of nested LetRec blocks
         relation.try_visit_mut_post(&mut |relation| -> Result<(), RecursionLimitError> {
-            // Disassemble `LetRec` into a `Let` stack if possible.
-            // If a `LetRec` remains, return the would-be `Let` bindings to it.
-            // This is to maintain `LetRec`-freedom for `LetRec`-free expressions.
-            let mut bindings = let_motion::harvest_non_recursive(relation);
-            if let MirRelationExpr::LetRec {
-                ids,
-                values,
-                limits,
-                body: _,
-            } = relation
-            {
-                bindings.extend(ids.drain(..).zip(values.drain(..).zip(limits.drain(..))));
-                support::replace_bindings_from_map(bindings, ids, values, limits);
-            } else {
-                for (id, (value, max_iter)) in bindings.into_iter().rev() {
-                    assert_none!(max_iter);
-                    *relation = MirRelationExpr::Let {
-                        id,
-                        value: Box::new(value),
-                        body: Box::new(relation.take_dangerous()),
-                    };
-                }
+            // Extract `Let` prefixes from `LetRec`, to reveal their non-recursive nature.
+            // This assists with hoisting e.g. arrangements out of `LetRec` blocks, a thing
+            // we don't promise to do, but it can be helpful to do. This also exposes more
+            // AST nodes to non-`LetRec` analyses, which don't always have parity with `LetRec`.
+            let bindings = let_motion::harvest_non_recursive(relation);
+            for (id, (value, max_iter)) in bindings.into_iter().rev() {
+                assert_none!(max_iter);
+                *relation = MirRelationExpr::Let {
+                    id,
+                    value: Box::new(value),
+                    body: Box::new(relation.take_dangerous()),
+                };
             }
 
             // Move a non-recursive suffix of bindings from the end of the LetRec
             // to the LetRec body.
+            // This is unsafe when applied to expressions which contain `ArrangeBy`,
+            // as if the extracted suffixes reference arrangements they will not be
+            // able to access those arrangements from outside the `LetRec` scope.
+            // It happens to work at the moment, so we don't touch it but should fix.
             let bindings = let_motion::harvest_nonrec_suffix(relation)?;
```

Comment:
When the new code above wraps a `LetRec` into a `Let`, this can make the `harvest_nonrec_suffix` call ineffective, because `relation` is no longer a `LetRec`, but `harvest_nonrec_suffix` just wants to match `LetRec` at the root. This seems to be happening in the output change in `normalize_lets.spec` below, where `l2` used to come out of the WMR, but now it remains in there (and similarly in `aoc_1204.slt` with `l7`, and in `aoc_1203.slt` with all of `l10` .. `l19` of the old version). Would it be possible to tweak the code to be able to still do `harvest_nonrec_suffix` in this case? It might be as simple as doing the `harvest_nonrec_suffix` call and the below code block not on `relation`, but on the innermost relation of the `Let` stack that is being assembled above?

Reply:
I'm still pondering, and I understand what you are saying, but what do you think about instead moving the re-assembly of let bindings extracted out the top until after we do this. I.e. make the logic
```rust
let top_bindings = let_motion::harvest_non_recursive(relation);
let bottom_bindings = let_motion:: harvest_nonrec_suffix(relation);
// reassemble
```

Reply:
Alternately, toggle the order we do things, to extract the suffix first, and then extract the prefix.

Reply:
PR updated to first extract suffixes, then extract prefixes.